rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandlerTest.groovy

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017 Schibsted ASA.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.rosco.providers
+
+import com.netflix.spinnaker.rosco.api.Artifact
+import com.netflix.spinnaker.rosco.api.Bake
+import com.netflix.spinnaker.rosco.api.BakeRequest
+import com.netflix.spinnaker.rosco.jobs.BakeRecipe
+import com.netflix.spinnaker.rosco.providers.util.TestDefaults
+import spock.lang.Specification
+import spock.lang.Subject
+
+class CloudProviderBakeHandlerTest extends Specification implements TestDefaults {
+
+  void 'we should create a fully decorated artifact if all base data is present'() {
+    setup:
+      def bakeRequest = new BakeRequest(build_info_url: SOME_BUILD_INFO_URL)
+      def bakeRecipe = new BakeRecipe(name: SOME_BAKE_RECIPE_NAME, version: SOME_APP_VERSION_STR, command: [])
+      def bakeDetails = new Bake(id: SOME_JOB_ID, ami: SOME_AMI_ID, image_name: SOME_IMAGE_NAME)
+      def cloudProvider = "docker"
+      def expectedArtifact = new Artifact (
+        name: bakeRecipe.name,
+        version: bakeRecipe.version,
+        type: cloudProvider,
+        reference: bakeDetails.ami,
+        metadata: [
+          build_info_url: bakeRequest.build_info_url,
+          build_number: bakeRequest.build_number
+        ]
+      )
+
+      @Subject
+      CloudProviderBakeHandler bakeHandler = Spy(CloudProviderBakeHandler)
+
+    when:
+      Artifact producedArtifact = bakeHandler.produceArtifactDecorationFrom(bakeRequest, bakeRecipe, bakeDetails, cloudProvider)
+
+    then:
+      producedArtifact == expectedArtifact
+  }
+
+  void 'we should not fail if data is missing for artifact decoration'() {
```

Comment:
Could break this down further and use a `where:` block to prove out each combination of missing data.